astropy/coordinates/baseframe.py

```diff
@@ -639,19 +645,60 @@ def __init__(self, *args, **kwargs):
             raise TypeError(
                 '{0}.__init__ had {1} remaining unhandled arguments'.format(
                     self.__class__.__name__, len(args)))
+
+        self._data = representation_data
+
+        values = {}
+        for fnm, fdefault in self.get_frame_attr_names().items():
+            # Read-only frame attributes are defined as FrameAttribue
+            # descriptors which are not settable, so set 'real' attributes as
+            # the name prefaced with an underscore.
+
+            if fnm in kwargs:
+                value = kwargs.pop(fnm)
+                setattr(self, '_' + fnm, value)
+                # Validate attribute by getting it.  If the instance has data,
+                # this also checks its shape is OK.  If not, we do it below.
+                values[fnm] = getattr(self, fnm)
+            else:
+                setattr(self, '_' + fnm, fdefault)
+                self._attr_names_with_defaults.append(fnm)
+
         if kwargs:
             raise TypeError(
                 'Coordinate frame got unexpected keywords: {0}'.format(
                     list(kwargs)))
 
-        self._data = representation_data
-
-        # We do ``is not None`` because self._data might evaluate to false for
+        # We do ``is None`` because self._data might evaluate to false for
         # empty arrays or data == 0
-        if self._data is not None:
+        if self._data is None:
+            # No data: we still need to check that any non-scalar attributes
+            # have consistent shapes.
+            shapes = {fnm: value.shape for fnm, value in values.items()
```

Comment:
It only requires the attribute to have a `shape` *if* it has a `size` (which is larger than 1). I added to the comment above to clarify this.